### PR TITLE
Milestone 2: Plutus transactions

### PR DESCRIPTION
**Pull request by PeterBenc**
Parsing plutus transaction includes showing 
- collateral return
- total collateral

We do not show datum hash, inline datum, inline datum, reference inputs, collateral inputs, redeemer or required singers

### CODE REVIEW

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+) => {
+  const amountRow = row(
+    'Amount',
+    text(assetValue(collateralReturnOutput.coin, ADA_TICKER)),
+  );
+  const accountIndex = collateralReturnOutput.isOwn
+    ? getAddressAccountIndex(collateralReturnOutput.ownAddress)
+    : undefined;
+
+  if (accountIndex) {
```
Comment (refi93) [inline packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:31]: as long as collateral is to be returned to user's own address, I wouldn't bother the user with that info as they just get the funds back. I think we already deal with change outputs in a similar way

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    return section([
+      heading(`Collateral return to ${getUiAccountIndex(accountIndex)}`),
```
Comment (refi93) [inline packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:33]:
```suggestion
      heading(`Collateral to be returned to ${getUiAccountIndex(accountIndex)}`),
```

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+      amountRow,
+    ]);
+  }
+
+  return section([
```
Comment (refi93) [inline packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:38]: As discussed offline, I'd consider "packing" advanced info like this into a single text field, if that's the only way to have "expandable" content in the metamask prompt

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    heading(`Collateral return `),
```
Comment (refi93) [inline packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:39]:
```suggestion
    heading(`Unspent collateral to be returned `),
```

**Comment by refi93** [comment]
I guess just the fixture doesn't reflect a real plutus tx, but I'd expect some `script` addresses to be involved in the tx as outputs, and also datum fields - shouldn't showing such info already be part of this PR?

Screenshot 2024-06-21 at 17 44 32

**Comment by refi93** [comment]
> We do not show datum hash, inline datum, inline datum, reference inputs, collateral inputs, redeemer or required singers

I'd expect that info to be visible to the user in a final version of the snap. Is this something we plan to deal with in a follow up PR/next milestone?

File: packages/cardano-metamask-snap/src/api/sdk/transaction.ts [hunk 5]
```diff
@@ -111,28 +123,41 @@ export const parseTransaction = ({
...
-  );
-
-  const collateral =
-    collateralReturn || totalCollateral
-      ? {
-          collateralReturn,
-          totalCollateral,
-        }
-      : undefined;
+  if (hasPlutusTxFields(parsedTransactionBody)) {
```
Comment (refi93) [inline packages/cardano-metamask-snap/src/api/sdk/transaction.ts:126]: just wondering - is this logic mirroring the way we detect plutus transactions in trezor connect? If so, I'd consider linking that here for reference

**Review by refi93 — APPROVED** [review]
minor comments/questions, othwerwise LGTM

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts (hunk 3 above)
Comment (PeterBenc) [reply packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:38]: Ok, if you agree, lets finish what is currently open, I will then open a new PR which adds expandable info

File: packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts (hunk 4 above)
Comment (PeterBenc) [reply packages/cardano-metamask-snap/src/api/cardano__signTransaction/ui/collateral.ts:39]: Yes I was following trezor UI here 
image

File: packages/cardano-metamask-snap/src/api/sdk/transaction.ts (hunk 5 above)
Comment (PeterBenc) [reply packages/cardano-metamask-snap/src/api/sdk/transaction.ts:126]: Yes, it follows how we in nufi determine if tx is a plutus one.
In nufi we reference ledger, we also do that here next to the definition of hasPlutusTxFields
https://github.com/cardano-foundation/ledgerjs-hw-app-cardano/blob/5e83052fd02c8e3bf88dbc837bb9219d80b41a40/src/parsing/transaction.ts#L341

**Comment by PeterBenc** [comment]
> > We do not show datum hash, inline datum, inline datum, reference inputs, collateral inputs, redeemer or required singers
> 
> I'd expect that info to be visible to the user in a final version of the snap. Is this something we plan to deal with in a follow up PR/next milestone?

Would be for showing that next to the outputs in the "main" UI, or down below next to raw tx and other tx details? We discussed this and decided to now show datum because trezor does not show it either.

File: packages/cardano-metamask-snap/src/api/sdk/transaction.ts (hunk 5 above)
Comment (refi93) [reply packages/cardano-metamask-snap/src/api/sdk/transaction.ts:126]: ok then I'd reference ledger here as well, for the sake of clarify

**Comment by PeterBenc** [comment]
Merged into develop

**Pull request closed without merging**